File: python/spider_browser/protocol/cdp_session.py
"""CDP JSON-RPC session over the Spider WebSocket transport.

The server proxies to Chrome's BROWSER-LEVEL CDP target.
We must discover/create a page target, attach with flatten=True,
then scope all Page/Runtime/Input commands with the sessionId.
"""

from __future__ import annotations

import asyncio
import json
from typing import Any

from .transport import Transport
from ..utils.errors import ProtocolError, TimeoutError
from ..utils.logger import logger

_CMD_TIMEOUT = 30.0


class CDPSession:
    def __init__(self, transport: Transport, *, command_timeout: float = _CMD_TIMEOUT) -> None:
        self._transport = transport
        self._next_id = 1
        self._pending: dict[int, asyncio.Future[dict[str, Any]]] = {}
        self._event_handlers: dict[str, list] = {}
        self._target_session_id: str | None = None
        self._timeout = command_timeout

    @property
    def session_id(self) -> str | None:
        return self._target_session_id

    # ------------------------------------------------------------------
    # Message routing
    # ------------------------------------------------------------------

    def handle_message(self, data: str) -> bool:
        try:
            msg = json.loads(data)
        except json.JSONDecodeError:
            return False

        # Response
        if isinstance(msg.get("id"), int):
            fut = self._pending.pop(msg["id"], None)
            if fut and not fut.done():
                fut.set_result(msg)
                return True
            return False

        # Event
        method = msg.get("method")
        if isinstance(method, str):
            params = msg.get("params", {})
            for h in list(self._event_handlers.get(method, [])):
                try:
                    h(params)
                except Exception:
                    pass
            for h in list(self._event_handlers.get("*", [])):
                try:
                    h({"method": method, **params})
                except Exception:
                    pass
            return True

        return False

    # ------------------------------------------------------------------
    # Command sending
    # ------------------------------------------------------------------

    async def send(self, method: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        """Send a browser-level CDP command."""
        cmd_id = self._next_id
        self._next_id += 1
        cmd: dict[str, Any] = {"id": cmd_id, "method": method, "params": params or {}}

        fut: asyncio.Future[dict[str, Any]] = asyncio.get_event_loop().create_future()
        self._pending[cmd_id] = fut
        await self._transport.send_async(json.dumps(cmd))
        try:
            return await asyncio.wait_for(fut, timeout=self._timeout)
        except asyncio.TimeoutError:
            self._pending.pop(cmd_id, None)
            raise TimeoutError(f"CDP command timeout: {method} ({self._timeout}s)")

    async def send_to_target(self, method: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        """Send a page-scoped CDP command (requires attachToPage first)."""
        if not self._target_session_id:
            raise ProtocolError("No target session — call attach_to_page() first")
        cmd_id = self._next_id
        self._next_id += 1
        cmd: dict[str, Any] = {
            "id": cmd_id,
            "method": method,
            "params": params or {},
            "sessionId": self._target_session_id,
        }
        fut: asyncio.Future[dict[str, Any]] = asyncio.get_event_loop().create_future()
        self._pending[cmd_id] = fut
        await self._transport.send_async(json.dumps(cmd))
        try:
            return await asyncio.wait_for(fut, timeout=self._timeout)
        except asyncio.TimeoutError:
            self._pending.pop(cmd_id, None)
            raise TimeoutError(f"CDP command timeout: {method} ({self._timeout}s)")

    def on(self, method: str, handler: Any) -> None:
        self._event_handlers.setdefault(method, []).append(handler)

    def off(self, method: str, handler: Any) -> None:
        try:
            self._event_handlers.get(method, []).remove(handler)
        except ValueError:
            pass

    # ------------------------------------------------------------------
    # Target management
    # ------------------------------------------------------------------

    async def attach_to_page(self) -> str:
        """Discover/create a page target, attach, enable domains."""
        await self.send("Target.setDiscoverTargets", {"discover": True})
        resp = await self.send("Target.getTargets")
        infos = (resp.get("result") or {}).get("targetInfos", [])

        target_id: str | None = None
        for t in infos:
            if t.get("type") == "page" and not (t.get("url") or "").startswith("devtools://"):
                target_id = t["targetId"]
                break

        if not target_id:
            cr = await self.send("Target.createTarget", {"url": "about:blank"})
            target_id = (cr.get("result") or {}).get("targetId")
            if not target_id:
                raise ProtocolError("Failed to create page target")

        attach_resp = await self.send("Target.attachToTarget", {"targetId": target_id, "flatten": True})
        sid = (attach_resp.get("result") or {}).get("sessionId")

        if not sid:
            event_fut: asyncio.Future[str] = asyncio.get_event_loop().create_future()

            def _handler(params: dict[str, Any]) -> None:
                s = params.get("sessionId")
                if s and not event_fut.done():
                    event_fut.set_result(s)

            self.on("Target.attachedToTarget", _handler)
            try:
                sid = await asyncio.wait_for(event_fut, timeout=5.0)
            except asyncio.TimeoutError:
                raise TimeoutError("Timeout waiting for Target.attachedToTarget")
            finally:
                self.off("Target.attachedToTarget", _handler)

        self._target_session_id = sid
        logger.info(f"attached to page target target_id={target_id} session_id={sid}")

        await self.send_to_target("Page.enable")
        await self.send_to_target("Runtime.enable")
        return sid

    # ------------------------------------------------------------------
    # High-level CDP commands
    # ------------------------------------------------------------------

    async def capture_screenshot(self) -> str:
        resp = await self.send_to_target("Page.captureScreenshot", {"format": "png"})
        data = (resp.get("result") or {}).get("data")
        if not isinstance(data, str):
            raise ProtocolError("captureScreenshot: missing result.data")
        return data

    async def get_html(self) -> str:
        resp = await self.send_to_target("Runtime.evaluate", {
            "expression": "document.documentElement.outerHTML",
            "returnByValue": True,
        })
        return self._extract_eval_value(resp)  # type: ignore

    async def evaluate(self, expression: str) -> Any:
        resp = await self.send_to_target("Runtime.evaluate", {
            "expression": expression,
            "returnByValue": True,
        })
        return self._extract_eval_value(resp)

    async def navigate(self, url: str) -> None:
        load_event = self._wait_for_event("Page.loadEventFired", 8.0)
        stop_event = self._wait_for_event("Page.frameStoppedLoading", 10.0)
        resp = await self.send_to_target("Page.navigate", {"url": url})
        error_text = (resp.get("result") or {}).get("errorText")
        if error_text:
            load_event.cancel()
            stop_event.cancel()
            from ..utils.errors import NavigationError
            if self._is_retryable_nav_error(error_text):
                raise NavigationError(f"Navigation failed: {error_text}")
            raise ProtocolError(f"Navigation failed: {error_text}")
        loaded = await load_event
        if not loaded:
            await stop_event

    async def navigate_fast(self, url: str) -> None:
        """Navigate with 5s max wait for load event."""
        load_event = self._wait_for_event("Page.loadEventFired", 5.0)
        resp = await self.send_to_target("Page.navigate", {"url": url})
        error_text = (resp.get("result") or {}).get("errorText")
        if error_text:
            load_event.cancel()
            from ..utils.errors import NavigationError
            if self._is_retryable_nav_error(error_text):
                raise NavigationError(f"Navigation failed: {error_text}")
            raise ProtocolError(f"Navigation failed: {error_text}")
        await load_event

    async def navigate_dom(self, url: str) -> None:
        """Navigate and return on DOMContentLoaded (3s max)."""
        dom_event = self._wait_for_event("Page.domContentEventFired", 3.0)
        resp = await self.send_to_target("Page.navigate", {"url": url})
        error_text = (resp.get("result") or {}).get("errorText")
        if error_text:
            dom_event.cancel()
            from ..utils.errors import NavigationError
            if self._is_retryable_nav_error(error_text):
                raise NavigationError(f"Navigation failed: {error_text}")
            raise ProtocolError(f"Navigation failed: {error_text}")
        await dom_event

    async def click_point(self, x: float, y: float) -> None:
        await self._mouse("mouseMoved", x, y, "none", 0)
        await self._mouse("mousePressed", x, y, "left", 1)
        await self._mouse("mouseReleased", x, y, "left", 1)

    async def right_click_point(self, x: float, y: float) -> None:
        await self._mouse("mouseMoved", x, y, "none", 0)
        await self._mouse("mousePressed", x, y, "right", 1)
        await self._mouse("mouseReleased", x, y, "right", 1)

    async def double_click_point(self, x: float, y: float) -> None:
        await self._mouse("mouseMoved", x, y, "none", 0)
        await self._mouse("mousePressed", x, y, "left", 1)
        await self._mouse("mouseReleased", x, y, "left", 1)
        await self._mouse("mousePressed", x, y, "left", 2)
        await self._mouse("mouseReleased", x, y, "left", 2)

    async def click_hold_point(self, x: float, y: float, hold_ms: int) -> None:
        await self._mouse("mouseMoved", x, y, "none", 0)
        await self._mouse("mousePressed", x, y, "left", 1)
        await asyncio.sleep(hold_ms / 1000)
        await self._mouse("mouseReleased", x, y, "left", 1)

    async def hover_point(self, x: float, y: float) -> None:
        await self._mouse("mouseMoved", x, y, "none", 0)

    async def drag_point(self, fx: float, fy: float, tx: float, ty: float) -> None:
        steps = 10
        await self._mouse("mouseMoved", fx, fy, "none", 0)
        await self._mouse("mousePressed", fx, fy, "left", 1)
        for i in range(1, steps + 1):
            t = i / steps
            await self._mouse("mouseMoved", fx + (tx - fx) * t, fy + (ty - fy) * t, "left", 0)
            await asyncio.sleep(0.016)
        await self._mouse("mouseReleased", tx, ty, "left", 1)

    async def insert_text(self, text: str) -> None:
        await self.send_to_target("Input.insertText", {"text": text})

    async def press_key(self, key: str, code: str, key_code: int) -> None:
        await self.send_to_target("Input.dispatchKeyEvent", {
            "type": "keyDown", "key": key, "code": code,
            "windowsVirtualKeyCode": key_code, "text": key,
        })
        await self.send_to_target("Input.dispatchKeyEvent", {
            "type": "keyUp", "key": key, "code": code,
            "windowsVirtualKeyCode": key_code,
        })

    async def key_down(self, key: str, code: str, key_code: int) -> None:
        await self.send_to_target("Input.dispatchKeyEvent", {
            "type": "keyDown", "key": key, "code": code,
            "windowsVirtualKeyCode": key_code, "text": key,
        })

    async def key_up(self, key: str, code: str, key_code: int) -> None:
        await self.send_to_target("Input.dispatchKeyEvent", {
            "type": "keyUp", "key": key, "code": code,
            "windowsVirtualKeyCode": key_code,
        })

    async def set_viewport(self, w: int, h: int, dpr: float = 2.0, mobile: bool = False) -> None:
        await self.send_to_target("Emulation.setDeviceMetricsOverride", {
            "width": w, "height": h, "deviceScaleFactor": dpr, "mobile": mobile,
        })

    def destroy(self) -> None:
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(Exception("session destroyed"))
        self._pending.clear()
        self._event_handlers.clear()

    # ------------------------------------------------------------------

    def _wait_for_event(self, method: str, timeout_s: float) -> asyncio.Task[bool]:
        """Wait for a CDP event with timeout. Returns a Task that resolves to True if event fired."""
        fut: asyncio.Future[bool] = asyncio.get_event_loop().create_future()

        def _handler(_params: Any) -> None:
            if not fut.done():
                fut.set_result(True)
            self.off(method, _handler)

        self.on(method, _handler)

        async def _wait() -> bool:
            try:
                return await asyncio.wait_for(asyncio.shield(fut), timeout=timeout_s)
            except asyncio.TimeoutError:
                self.off(method, _handler)
                if not fut.done():
                    fut.set_result(False)
                return False
            except asyncio.CancelledError:
                self.off(method, _handler)
                return False

        return asyncio.get_event_loop().create_task(_wait())

    @staticmethod
    def _is_retryable_nav_error(error_text: str) -> bool:
        retryable = [
            "ERR_ABORTED", "ERR_CONNECTION_RESET", "ERR_CONNECTION_CLOSED",
            "ERR_EMPTY_RESPONSE", "ERR_BLOCKED_BY_CLIENT",
            "ERR_SSL_PROTOCOL_ERROR", "ERR_SSL_VERSION_OR_CIPHER_MISMATCH",
        ]
        return any(e in error_text for e in retryable)

    async def _mouse(self, typ: str, x: float, y: float, button: str, click: int) -> None:
        await self.send_to_target("Input.dispatchMouseEvent", {
            "type": typ, "x": x, "y": y, "button": button, "clickCount": click,
        })

    def _extract_eval_value(self, resp: dict[str, Any]) -> Any:
        err = resp.get("error")
        if err:
            raise ProtocolError(f"CDP error: {err.get('message', err)}")
        r = resp.get("result", {})
        return r.get("result", {}).get("value")

File: typescript/protocol/cdp-session.ts
import type { Transport } from './transport.js';
import type { CDPResponse, CDPEvent } from './types.js';
import { NavigationError, ProtocolError, TimeoutError } from '../utils/errors.js';
import { logger } from '../utils/logger.js';

type PendingResolver = {
  resolve: (value: CDPResponse) => void;
  reject: (err: Error) => void;
  timer: ReturnType<typeof setTimeout>;
};

type EventHandler = (params: Record<string, unknown>) => void;

/**
 * CDP JSON-RPC session over the Spider WebSocket transport.
 *
 * The server proxies to Chrome's BROWSER-LEVEL CDP target.
 * This means we must:
 *   1. Discover/create a page target
 *   2. Attach to it with flatten: true to get a sessionId
 *   3. Send all Page/Runtime/Input commands with that sessionId
 *
 * Used for Chrome, Chrome-H, Servo, and LightPanda browsers.
 */
export class CDPSession {
  private nextId = 1;
  private pending = new Map<number, PendingResolver>();
  private eventHandlers = new Map<string, Set<EventHandler>>();
  private transport: Transport;
  /** CDP session ID from Target.attachToTarget (required for page-level commands). */
  private targetSessionId: string | undefined;
  private commandTimeoutMs: number;

  constructor(transport: Transport, opts?: { commandTimeoutMs?: number }) {
    this.transport = transport;
    this.commandTimeoutMs = opts?.commandTimeoutMs ?? 30_000;
  }

  /** Get the attached target session ID. */
  get sessionId(): string | undefined {
    return this.targetSessionId;
  }

  /** Process a raw message from the transport. Returns true if handled. */
  handleMessage(data: string): boolean {
    let msg: any;
    try {
      msg = JSON.parse(data);
    } catch {
      return false;
    }

    // Response to a command (has "id")
    if (typeof msg.id === 'number') {
      const pending = this.pending.get(msg.id);
      if (pending) {
        this.pending.delete(msg.id);
        clearTimeout(pending.timer);
        pending.resolve(msg as CDPResponse);
        return true;
      }
      return false;
    }

    // Event (has "method", no "id")
    if (typeof msg.method === 'string') {
      const handlers = this.eventHandlers.get(msg.method);
      if (handlers) {
        const params = msg.params ?? {};
        for (const h of handlers) {
          try {
            h(params);
          } catch {
            // ignore handler errors
          }
        }
      }
      // Also fire wildcard handlers (for target discovery)
      const wildcardHandlers = this.eventHandlers.get('*');
      if (wildcardHandlers) {
        for (const h of wildcardHandlers) {
          try {
            h({ method: msg.method, ...(msg.params ?? {}) });
          } catch {
            // ignore
          }
        }
      }
      return true;
    }

    return false;
  }

  /** Send a CDP command and wait for the response. */
  async send(method: string, params?: Record<string, unknown>): Promise<CDPResponse> {
    const id = this.nextId++;
    const cmd: any = { id, method, params: params ?? {} };

    return new Promise<CDPResponse>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new TimeoutError(`CDP command timeout: ${method} (${this.commandTimeoutMs}ms)`));
      }, this.commandTimeoutMs);

      this.pending.set(id, { resolve, reject, timer });
      try {
        this.transport.send(JSON.stringify(cmd));
      } catch (err) {
        this.pending.delete(id);
        clearTimeout(timer);
        reject(err);
      }
    });
  }

  /**
   * Send a CDP command scoped to the attached page session.
   * This is what you use for Page.*, Runtime.*, Input.* commands.
   */
  async sendToTarget(method: string, params?: Record<string, unknown>): Promise<CDPResponse> {
    if (!this.targetSessionId) {
      throw new ProtocolError('No target session — call attachToPage() first');
    }
    const id = this.nextId++;
    const cmd: any = {
      id,
      method,
      params: params ?? {},
      sessionId: this.targetSessionId,
    };

    return new Promise<CDPResponse>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new TimeoutError(`CDP command timeout: ${method} (${this.commandTimeoutMs}ms)`));
      }, this.commandTimeoutMs);

      this.pending.set(id, { resolve, reject, timer });
      try {
        this.transport.send(JSON.stringify(cmd));
      } catch (err) {
        this.pending.delete(id);
        clearTimeout(timer);
        reject(err);
      }
    });
  }

  /** Subscribe to a CDP event. */
  on(method: string, handler: EventHandler): void {
    let set = this.eventHandlers.get(method);
    if (!set) {
      set = new Set();
      this.eventHandlers.set(method, set);
    }
    set.add(handler);
  }

  /** Unsubscribe from a CDP event. */
  off(method: string, handler: EventHandler): void {
    this.eventHandlers.get(method)?.delete(handler);
  }

  // -------------------------------------------------------------------
  // Browser-level target management
  // -------------------------------------------------------------------

  /**
   * Discover page targets, find or create one, attach to it, and enable
   * the required CDP domains (Page, Runtime).
   *
   * This is the key initialization step — without this, Page.navigate and
   * Runtime.evaluate won't work because we're connected at browser level.
   */
  async attachToPage(): Promise<string> {
    // Step 1: Enable target discovery
    await this.send('Target.setDiscoverTargets', { discover: true });

    // Step 2: Always create a fresh page target.
    // When multiple sessions share a Chrome instance, reusing existing page targets
    // causes content contamination (wrong cookies, storage, cached content).
    // The server injects browserContextId into Target.createTarget to isolate
    // each session's pages in their own browser context.
    let pageTargetId: string | undefined;
    logger.debug('creating fresh page target for session isolation');
    const createResp = await this.send('Target.createTarget', { url: 'about:blank' });
    pageTargetId = (createResp.result as any)?.targetId;
    if (!pageTargetId) {
      throw new ProtocolError('Failed to create page target');
    }
    logger.debug(`created page target: ${pageTargetId}`);

    // Step 4: Attach to the target with flatten: true
    // flatten: true gives us a flat session — commands with sessionId go directly to the target
    const attachResp = await this.send('Target.attachToTarget', {
      targetId: pageTargetId,
      flatten: true,
    });

    // The sessionId comes either from the response or from a Target.attachedToTarget event
    let sessionId = (attachResp.result as any)?.sessionId;

    if (!sessionId) {
      // Some Chrome versions send the sessionId via an event instead
      // Wait briefly for the event
      sessionId = await new Promise<string>((resolve, reject) => {
        const timeout = setTimeout(() => {
          this.off('Target.attachedToTarget', handler);
          reject(new TimeoutError('Timeout waiting for Target.attachedToTarget event'));
        }, 5000);

        const handler = (params: Record<string, unknown>) => {
          const sid = (params as any).sessionId;
          if (sid) {
            clearTimeout(timeout);
            this.off('Target.attachedToTarget', handler);
            resolve(sid);
          }
        };
        this.on('Target.attachedToTarget', handler);
      });
    }

    this.targetSessionId = sessionId;
    logger.info(`attached to page target`, { targetId: pageTargetId, sessionId });

    // Step 5: Enable domains on the target session
    await this.sendToTarget('Page.enable');
    await this.sendToTarget('Runtime.enable');

    return sessionId;
  }

  // -------------------------------------------------------------------
  // High-level CDP commands (all use sendToTarget for page-scoped ops)
  // -------------------------------------------------------------------

  /** Capture a screenshot as base64 PNG (10s timeout to avoid blocking on heavy pages). */
  async captureScreenshot(): Promise<string> {
    const screenshotTimeout = Math.min(this.commandTimeoutMs, 10_000);
    const id = this.nextId++;
    const cmd: any = {
      id,
      method: 'Page.captureScreenshot',
      params: { format: 'png' },
      sessionId: this.targetSessionId,
    };

    const resp = await new Promise<CDPResponse>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new TimeoutError(`CDP command timeout: Page.captureScreenshot (${screenshotTimeout}ms)`));
      }, screenshotTimeout);

      this.pending.set(id, { resolve, reject, timer });
      try {
        this.transport.send(JSON.stringify(cmd));
      } catch (err) {
        this.pending.delete(id);
        clearTimeout(timer);
        reject(err);
      }
    });

    const data = resp.result?.['data'];
    if (typeof data !== 'string') {
      throw new ProtocolError('captureScreenshot: missing result.data');
    }
    return data;
  }

  /** Get full page HTML. */
  async getHTML(): Promise<string> {
    const resp = await this.sendToTarget('Runtime.evaluate', {
      expression: 'document.documentElement.outerHTML',
      returnByValue: true,
    });
    return this.extractEvalValue(resp) as string;
  }

  /** Evaluate a JavaScript expression and return the value. */
  async evaluate(expression: string): Promise<unknown> {
    const resp = await this.sendToTarget('Runtime.evaluate', {
      expression,
      returnByValue: true,
    });
    return this.extractEvalValue(resp);
  }

  /** Wait for a CDP event to fire, with timeout. Returns true if event fired, false on timeout. */
  private waitForEvent(method: string, timeoutMs: number): Promise<boolean> {
    return new Promise<boolean>((resolve) => {
      const timer = setTimeout(() => {
        this.off(method, handler);
        resolve(false);
      }, timeoutMs);

      const handler = () => {
        clearTimeout(timer);
        this.off(method, handler);
        resolve(true);
      };
      this.on(method, handler);
    });
  }

  /** Navigate to a URL and wait for the page to load.
   *
   * Uses short timeouts (8s load + 10s fallback = 10s max after load timeout) because:
   * - Most pages fire loadEventFired in <5s
   * - SPAs that don't fire it in 8s rarely will (they use client-side rendering)
   * - content() has built-in SPA polling that handles incomplete loads gracefully
   * - Short navigate means more page timeout budget for retries and content extraction
   */
  async navigate(url: string): Promise<void> {
    // Start listening for load events BEFORE sending navigate
    const loadPromise = this.waitForEvent('Page.loadEventFired', 8_000);
    const stopPromise = this.waitForEvent('Page.frameStoppedLoading', 10_000);

    const resp = await this.sendToTarget('Page.navigate', { url });

    // Check for navigation errors (e.g. net::ERR_NAME_NOT_RESOLVED)
    const errorText = (resp.result as any)?.errorText;
    if (errorText) {
      // Clean up event listeners
      loadPromise.catch(() => {});
      stopPromise.catch(() => {});
      // Retryable nav errors: close connection and try fresh
      if (isRetryableNavError(errorText)) {
        throw new NavigationError(`Navigation failed: ${errorText}`);
      }
      throw new ProtocolError(`Navigation failed: ${errorText}`);
    }

    // Wait for load event; if it times out, fall back to frameStoppedLoading
    const loaded = await loadPromise;
    if (!loaded) {
      // loadEventFired timed out — wait briefly for frameStoppedLoading
      await stopPromise;
      // Proceed silently even if both timed out — content() polls for content
    }
  }

  /**
   * Navigate without waiting for full page load — returns after the navigation
   * response is received and a brief initial load wait (5s).
   * Use with contentWithEarlyReturn() for SPAs that never fire loadEventFired.
   */
  async navigateFast(url: string): Promise<void> {
    const loadPromise = this.waitForEvent('Page.loadEventFired', 5_000);

    const resp = await this.sendToTarget('Page.navigate', { url });

    const errorText = (resp.result as any)?.errorText;
    if (errorText) {
      loadPromise.catch(() => {});
      if (isRetryableNavError(errorText)) {
        throw new NavigationError(`Navigation failed: ${errorText}`);
      }
      throw new ProtocolError(`Navigation failed: ${errorText}`);
    }

    // Wait up to 5s for load — if it fires, great. If not, proceed to polling.
    await loadPromise;
  }

  /**
   * Navigate and return as soon as DOMContentLoaded fires (3s max).
   * Much faster than navigate() for SPAs — the DOM shell is ready at
   * DOMContentLoaded, and content() polling handles async rendering.
   * Falls through silently if DOMContentLoaded doesn't fire.
   */
  async navigateDom(url: string): Promise<void> {
    const domPromise = this.waitForEvent('Page.domContentEventFired', 3_000);

    const resp = await this.sendToTarget('Page.navigate', { url });

    const errorText = (resp.result as any)?.errorText;
    if (errorText) {
      domPromise.catch(() => {});
      if (isRetryableNavError(errorText)) {
        throw new NavigationError(`Navigation failed: ${errorText}`);
      }
      throw new ProtocolError(`Navigation failed: ${errorText}`);
    }

    await domPromise;
  }

  /** Dispatch a mouse event. */
  async dispatchMouseEvent(
    type: string,
    x: number,
    y: number,
    button: string = 'none',
    clickCount: number = 0,
  ): Promise<void> {
    await this.sendToTarget('Input.dispatchMouseEvent', { type, x, y, button, clickCount });
  }

  /** Click at coordinates (mouseMoved -> mousePressed -> mouseReleased). */
  async clickPoint(x: number, y: number): Promise<void> {
    await this.dispatchMouseEvent('mouseMoved', x, y, 'none', 0);
    await this.dispatchMouseEvent('mousePressed', x, y, 'left', 1);
    await this.dispatchMouseEvent('mouseReleased', x, y, 'left', 1);
  }

  /** Right-click at coordinates. */
  async rightClickPoint(x: number, y: number): Promise<void> {
    await this.dispatchMouseEvent('mouseMoved', x, y, 'none', 0);
    await this.dispatchMouseEvent('mousePressed', x, y, 'right', 1);
    await this.dispatchMouseEvent('mouseReleased', x, y, 'right', 1);
  }

  /** Double-click at coordinates. */
  async doubleClickPoint(x: number, y: number): Promise<void> {
    await this.dispatchMouseEvent('mouseMoved', x, y, 'none', 0);
    await this.dispatchMouseEvent('mousePressed', x, y, 'left', 1);
    await this.dispatchMouseEvent('mouseReleased', x, y, 'left', 1);
    await this.dispatchMouseEvent('mousePressed', x, y, 'left', 2);
    await this.dispatchMouseEvent('mouseReleased', x, y, 'left', 2);
  }

  /** Click and hold at coordinates for a duration. */
  async clickHoldPoint(x: number, y: number, holdMs: number): Promise<void> {
    await this.dispatchMouseEvent('mouseMoved', x, y, 'none', 0);
    await this.dispatchMouseEvent('mousePressed', x, y, 'left', 1);
    await sleep(holdMs);
    await this.dispatchMouseEvent('mouseReleased', x, y, 'left', 1);
  }

  /** Hover (mouseMoved only). */
  async hoverPoint(x: number, y: number): Promise<void> {
    await this.dispatchMouseEvent('mouseMoved', x, y, 'none', 0);
  }

  /** Smooth drag from point to point (10-step interpolation). */
  async dragPoint(fromX: number, fromY: number, toX: number, toY: number): Promise<void> {
    const steps = 10;
    await this.dispatchMouseEvent('mouseMoved', fromX, fromY, 'none', 0);
    await this.dispatchMouseEvent('mousePressed', fromX, fromY, 'left', 1);
    for (let i = 1; i <= steps; i++) {
      const t = i / steps;
      const x = fromX + (toX - fromX) * t;
      const y = fromY + (toY - fromY) * t;
      await this.dispatchMouseEvent('mouseMoved', x, y, 'left', 0);
      await sleep(16);
    }
    await this.dispatchMouseEvent('mouseReleased', toX, toY, 'left', 1);
  }

  /** Insert text via Input.insertText. */
  async insertText(text: string): Promise<void> {
    await this.sendToTarget('Input.insertText', { text });
  }

  /** Press a key (keyDown + keyUp). */
  async pressKey(key: string, code: string, keyCode: number): Promise<void> {
    await this.sendToTarget('Input.dispatchKeyEvent', {
      type: 'keyDown',
      key,
      code,
      windowsVirtualKeyCode: keyCode,
      text: key,
    });
    await this.sendToTarget('Input.dispatchKeyEvent', {
      type: 'keyUp',
      key,
      code,
      windowsVirtualKeyCode: keyCode,
    });
  }

  /** Send keyDown event. */
  async keyDown(key: string, code: string, keyCode: number): Promise<void> {
    await this.sendToTarget('Input.dispatchKeyEvent', {
      type: 'keyDown',
      key,
      code,
      windowsVirtualKeyCode: keyCode,
      text: key,
    });
  }

  /** Send keyUp event. */
  async keyUp(key: string, code: string, keyCode: number): Promise<void> {
    await this.sendToTarget('Input.dispatchKeyEvent', {
      type: 'keyUp',
      key,
      code,
      windowsVirtualKeyCode: keyCode,
    });
  }

  /** Set viewport via Emulation.setDeviceMetricsOverride. */
  async setViewport(
    width: number,
    height: number,
    deviceScaleFactor: number = 2,
    mobile: boolean = false,
  ): Promise<void> {
    await this.sendToTarget('Emulation.setDeviceMetricsOverride', {
      width,
      height,
      deviceScaleFactor,
      mobile,
    });
  }

  /** Clean up all pending commands and event handlers. Rejects in-flight commands. */
  destroy(): void {
    // Snapshot and clear first to prevent re-entrant issues from reject callbacks
    const pendingSnapshot = [...this.pending.values()];
    this.pending.clear();
    this.eventHandlers.clear();
    this.targetSessionId = undefined;

    for (const pending of pendingSnapshot) {
      clearTimeout(pending.timer);
      pending.reject(new Error('session destroyed'));
    }
  }

  private extractEvalValue(resp: CDPResponse): unknown {
    if (resp.error) {
      throw new ProtocolError(`CDP error: ${resp.error.message}`);
    }
    const result = resp.result as any;
    return result?.result?.value;
  }
}

/** Navigation errors that are retryable by closing the connection and reconnecting. */
const RETRYABLE_NAV_ERRORS = [
  'net::ERR_ABORTED',
  'net::ERR_CONNECTION_RESET',
  'net::ERR_CONNECTION_CLOSED',
  'net::ERR_CONNECTION_REFUSED',
  'net::ERR_CONNECTION_TIMED_OUT',
  'net::ERR_TIMED_OUT',
  'net::ERR_EMPTY_RESPONSE',
  'net::ERR_SOCKET_NOT_CONNECTED',
  'net::ERR_NETWORK_CHANGED',
];

function isRetryableNavError(errorText: string): boolean {
  return RETRYABLE_NAV_ERRORS.some((e) => errorText.includes(e));
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
